### PR TITLE
Enum serialized as string

### LINKED ISSUE #19314: Enum support for Compact Serializer
Currently when object serialized has enum fields (or it's an enum itself), an error is thrown:

> NoSuchMethodException: com.myexample.MyEnum.<init>()

A custom serializer must be made for each possible enum, which is not very user-friendly. Nice enums are core Java features, it would be great to make them serializable by default.

### PR DESCRIPTION
**Pull request by TomaszGaweda**
Fixes #19314 

Adds serialization of enum types.

Breaking changes (list specific methods/types/messages):
* serialized form: format of Enums in Compact serializer

Checklist:
- [ ] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [ ] Add `Add to Release Notes` label if changes should be mentioned in release notes or `Not Release Notes content` if changes are not relevant for release notes
- [ ] Request reviewers if possible
- [ ] New public APIs have `@Nonnull/@Nullable` annotations
- [ ] New public APIs have `@since` tags in Javadoc
- [ ] Send backports/forwardports if fix needs to be applied to past/future releases

### CODE REVIEW

**Comment by TomaszGaweda** [comment]
Hi @mdumandag, can you please review once you have some time? Thanks in advance!

File: hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/compact/CompactStreamSerializerTest.java [hunk 1]
```diff
@@ -390,25 +387,28 @@ public void testFieldOrder() throws IOException {
...
         ReflectiveCompactSerializer reflectiveCompactSerializer = new ReflectiveCompactSerializer();
-        EmployerDTO employerDTO = new EmployerDTO("nbss", 40, ids, employeeDTO, employeeDTOS);
+        EmployerDTO employerDTO = new EmployerDTO("nbss", 40, HIRING, ids, employeeDTO, employeeDTOS);
         reflectiveCompactSerializer.write(writer, employerDTO);
 
         Schema schema = writer.build();
 
-        assertEquals(schema.getField("zcode").getOffset(), 0);
-        assertEquals(schema.getField("zcode").getIndex(), -1);
+        assertEquals(0, schema.getField("zcode").getOffset());
```
Comment (TomaszGaweda) [inline hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/compact/CompactStreamSerializerTest.java:395]: expected and actual were swapped, just changed order to correct one

File: hazelcast/src/main/java/com/hazelcast/nio/serialization/compact/CompactReader.java [hunk 2]
```diff
@@ -376,6 +376,19 @@
...
+     *
+     * @param fieldName name of the field.
+     * @param enumType underlying enum type.
+     * @return the value of the field.
+     * @throws HazelcastSerializationException if the field does not exist in the schema
+     *                                         or the type of the field does not match
+     *                                         with the one defined in the schema.
+     */
+    @Nullable
+    <T extends Enum> T readEnum(@Nonnull String fieldName, Class<T> enumType);
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/nio/serialization/compact/CompactReader.java:390]: We decided to not give an API for enums/enum arrays in both the GenericRecord and CompactReader/Writer interfaces. For the CompactReader/Writer, we want to maintain a FieldType for each kind of object we support.

For the GenericRecord, there are more reasons:
- We want GenericRecord API to be free of any type of class as on the server-side, there might not be such class. After all, GenericRecord is designed to be used as a way to represent domain objects without any sort of classes.
- We also want to maintain FieldType/object mapping there.
- Also, we want GenericRecord to be a polyglot. You may represent enums with Strings in Java, but they might be integers in C++.We want to maintain consistency across different languages.

Therefore, could you please remove readEnum and readEnumArray from the CompactReader and writeEnum and writeEnumArray from the CompactWriter?

We just want to implicitly support enums/enum arrays for the reflective serializer. For the explicit serializer use cases (where you use CompactReader/Writer), we want them to represent enums in the way they want.

Comment (TomaszGaweda) [reply hazelcast/src/main/java/com/hazelcast/nio/serialization/compact/CompactReader.java:390]: Ach, I get it, "inline" this in serializer. Ok, will be done soon, sorry for misunderstanding

Comment (TomaszGaweda) [reply hazelcast/src/main/java/com/hazelcast/nio/serialization/compact/CompactReader.java]: Done

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/ReflectiveCompactSerializer.java [hunk 3]
```diff
@@ -39,36 +39,8 @@
...
-import static com.hazelcast.nio.serialization.FieldType.SHORT_ARRAY;
-import static com.hazelcast.nio.serialization.FieldType.TIME;
-import static com.hazelcast.nio.serialization.FieldType.TIMESTAMP;
-import static com.hazelcast.nio.serialization.FieldType.TIMESTAMP_ARRAY;
-import static com.hazelcast.nio.serialization.FieldType.TIMESTAMP_WITH_TIMEZONE;
-import static com.hazelcast.nio.serialization.FieldType.TIMESTAMP_WITH_TIMEZONE_ARRAY;
-import static com.hazelcast.nio.serialization.FieldType.TIME_ARRAY;
-import static com.hazelcast.nio.serialization.FieldType.UTF;
-import static com.hazelcast.nio.serialization.FieldType.UTF_ARRAY;
+import static com.hazelcast.nio.serialization.FieldType.*;
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/ReflectiveCompactSerializer.java:42]: I assume star import will annoy the checkstyle

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/ReflectiveCompactSerializer.java [hunk 4]
```diff
@@ -287,6 +261,14 @@ private void createFastReadWriteCaches(Class clazz) throws IOException {
                     }
                 };
                 writers[index] = (w, o) -> w.writeTimestampWithTimezone(name, (OffsetDateTime) field.get(o));
+            } else if (type.isEnum()) {
+                readers[index] = (reader, schema, o) -> {
+                    if (fieldExists(schema, name, UTF)) {
+                        String enumName = reader.readString(name);
+                        field.set(o, Enum.valueOf((Class<? extends Enum>) type, enumName));
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/ReflectiveCompactSerializer.java:268]: enumName can be null if the enum field is null. I think we should handle this case and set the field to null if the enumName is null

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/ReflectiveCompactSerializer.java [hunk 5]
```diff
@@ -287,6 +261,14 @@ private void createFastReadWriteCaches(Class clazz) throws IOException {
...
+                    }
+                };
+                writers[index] = (w, o) -> w.writeString(name, ((Enum) field.get(o)).name());
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/ReflectiveCompactSerializer.java:271]: Similar comment as above. field.get(o) might return null. We could directly write null to writeString

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/ReflectiveCompactSerializer.java [hunk 6]
```diff
@@ -387,6 +369,31 @@ private void createFastReadWriteCaches(Class clazz) throws IOException {
...
                     writers[index] = (w, o) -> w.writeTimestampWithTimezoneArray(name, (OffsetDateTime[]) field.get(o));
+                } else if (componentType.isEnum()) {
+                    readers[index] = (reader, schema, o) -> {
+                        if (fieldExists(schema, name, UTF_ARRAY)) {
+                            String[] stringArray = reader.readStringArray(name);
+                            Enum[] enumArray = null;
+                            if (stringArray != null) {
+                                enumArray = new Enum[stringArray.length];
+                                for (int i = 0; i < stringArray.length; i++) {
+                                    enumArray[i] = Enum.valueOf((Class<? extends Enum>) componentType, stringArray[i]);
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/ReflectiveCompactSerializer.java:380]: Similar comment as above

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/ReflectiveCompactSerializer.java [hunk 7]
```diff
@@ -387,6 +369,31 @@ private void createFastReadWriteCaches(Class clazz) throws IOException {
...
+                            field.set(o, enumArray);
+                        }
+                    };
+                    writers[index] = (w, o) -> {
+                        Enum[] values = (Enum[]) field.get(o);
+                        String[] stringArray = null;
+                        if (values != null) {
+                            stringArray = new String[values.length];
+                            for (int i = 0; i < values.length; i++) {
+                                stringArray[i] = values[i].name();
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/ReflectiveCompactSerializer.java:392]: Similar comment as above

File: hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/compact/CompactStreamSerializerTest.java [hunk 8]
```diff
@@ -32,13 +32,7 @@
...
 import com.hazelcast.test.annotation.ParallelJVMTest;
 import com.hazelcast.test.annotation.QuickTest;
-import example.serialization.BitsDTO;
-import example.serialization.EmployeeDTO;
-import example.serialization.EmployeeDTOSerializer;
-import example.serialization.EmployeeWithSerializerDTO;
-import example.serialization.EmployerDTO;
-import example.serialization.ExternalizableEmployeeDTO;
-import example.serialization.NodeDTO;
+import example.serialization.*;
```
Comment (mdumandag) [inline hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/compact/CompactStreamSerializerTest.java:35]: Let's revert star import to make checkstyle happy

File: hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/ReflectiveCompactSerializer.java (hunk 4 above)
Comment (TomaszGaweda) [reply hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/ReflectiveCompactSerializer.java]: good catch, changed

File: hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/compact/CompactStreamSerializerTest.java (hunk 8 above)
Comment (TomaszGaweda) [reply hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/compact/CompactStreamSerializerTest.java]: Done

**Comment by mdumandag** [comment]
run-lab-run

**Comment by mdumandag** [comment]
@TomaszGaweda Can you check the checkstyle errors reported by the PR builder https://github.com/hazelcast/hazelcast/pull/19488#issuecomment-917947346?

**Comment by TomaszGaweda** [comment]
> 
> 
> @TomaszGaweda Can you check the checkstyle errors reported by the PR builder [#19488 (comment)](https://github.com/hazelcast/hazelcast/pull/19488#issuecomment-917947346)?

Sure, done. 
BTW. Checkstyle reports few other issues in not affected files; I didn't notice those errors because of that. WDYT of some general checkstyle repair PR?

**Comment by mdumandag** [comment]
run-lab-run

**Comment by mdumandag** [comment]
@TomaszGaweda Sure, just make sure that you rebased your code with the recent changes. Maybe they are fixed in the master. I see that the latest commit passed the checks without a problem. https://github.com/hazelcast/hazelcast/runs/3567122967

**Comment by mdumandag** [comment]
run-lab-run

**Review by mdumandag — APPROVED** [review]
Thanks for the changes @TomaszGaweda. Looks good to me. Let me find a second reviewer

**Review by sancar — APPROVED** [review]

**Comment by mdumandag** [comment]
@TomaszGaweda, this PR is ready to merge, but we will have to postpone merging it for some time.

Right now, we are very close to the code freeze date for 5.0, and we only allow stabilization-related PRs to merge. 

Once the 5.0 is out, I will make sure to merge this PR. Thanks for your contribution & patience.

**Review by ihsandemir — APPROVED** [review]

**Comment by TomaszGaweda** [comment]
Understandable @mdumandag, thanks for your review and patience :)

**Pull request merged**